src/CodeGen_LLVM.h

```diff
@@ -84,6 +84,14 @@ class CodeGen_LLVM : public IRVisitor {
     virtual void compile_buffer(const Buffer &buffer);
     // @}
 
+    /** Helper functions for compiling Halide functions to llvm
```

Comment:
I just noticed you mentioned that compile_func should be in between, but this really isn't the case. compile_func calls begin_func and end_func itself. This refactoring is to make it so one can override compile_func, calling begin_func and end_func themselves, generating the function body in between.

Thinking about this more, begin_func and end_func shouldn't take lowered func, they should just take the argument list (and function name). I'm going to do that now.